### ISSUE TITLE
ContractFactory violates Liskov substitution principle 

### ISSUE THREAD

**Issue by sz-piotr** [ethereum-ts/TypeChain#215]
Suppose i have a simple contract
```
pragma solidity ^0.6.0;

contract MyContract {
    constructor (address a) public {
    }
}
```

Using typechain i obtain `MyContractFactory`.

Unfortunately `MyContractFactory` lacks functionality compared to the ethers `ContractFactory`.
1. new MyContractFactory(wallet).deploy(...) does not accept overrides as the last argument
2. casting new MyContractFactory(wallet) as ContractFactory and then calling deploy with an incorrect number of arguments we do not get the proper error about argument count from ethers.

Suggested fix (not tested):

```ts
class MyContractFactory {
  deploy(a: string, overrides: providers.TransactionRequest): Promise<MyContract>
  deploy(...args: any[]): Promise<MyContract> {
    super.deploy(...args) as Promise<MyContract>
  }
}
```

**Comment by krzkaczor** [comment]
CC: @quezak

**Comment by quezak** [comment]
@sz-piotr can you point out where does it say the original `ContractFactory` does accept transaction overrides? I don't see that it does it in the [docs](https://docs.ethers.io/ethers.js/html/api-contract.html#deployment) nor in its [typings](https://github.com/ethers-io/ethers.js/blob/master/contract.d.ts#L81).

**Comment by sz-piotr** [comment]
@quezak Sure. Here is the relevant code:
https://github.com/ethers-io/ethers.js/blob/748f89660ac4bf60175f936da755cf1a0d370f46/src.ts/contract.ts#L798

**Comment by quezak** [comment]
Ok, nice hidden feature, gratz to them on not mentioning it anywhere else :joy: 
I'll have to update the typings then.

**Comment by krzkaczor** [comment]
📦 Released as `typechain-target-ethers@1.0.4`
